File: src/tja/scrapers/bailii.nim
import std/[re, strutils, times]
import chronicles
from ../../tja/base_scraper import BaseScraper, initBaseScraper, makeRequest
from ../../tja/data_models import CaseData, initCaseData
from ../../tja/helpers import sanitizeText, normalizeCourtName


type
  BAILIIScraper* = ref object of BaseScraper

method baseUrl*(self: BAILIIScraper): string = "https://www.bailii.org"
method jurisdiction*(self: BAILIIScraper): string = "United Kingdom"

proc newBAILIIScraper*(): BAILIIScraper = BAILIIScraper(initBaseScraper())

proc searchCases*(self: BAILIIScraper; query: string; startDate = ""; endDate = ""; court = ""; limit = 100; jurisdiction = "uk"): seq[CaseData] =
  var params: seq[(string,string)] = @[("method","boolean"), ("results", $(min(limit, 200)))]
  if query.len > 0: params.add ("query", query)
  if startDate.len > 0: params.add ("date_from", startDate)
  if endDate.len > 0: params.add ("date_to", endDate)
  if court.len > 0: params.add ("court", court)
  if jurisdiction in ["uk","ie","ni","scot"]: params.add ("jurisdiction", jurisdiction)
  let url = self.baseUrl() & "/cgi-bin/markup.cgi"
  let resp = self.makeRequest(url, params=params)
  var out: seq[CaseData] = @[]
  let linkRe = re"<a[^>]+href=\"(/(uk|ie|ni|scot)/cases/[^\"]+)\"[^>]*>(.*?)</a>"
  for m in findAll(resp.body, linkRe):
    var c = initCaseData()
    c.caseName = sanitizeText(m.group(0,3))
    let href = m.group(0,1)
    c.url = if href.startsWith("http"): href else: self.baseUrl() & href
    let idRe = re"/(uk|ie|ni|scot)/cases/([^/]+/\d+/\d+)"
    let mm = idRe.find(c.url)
    if mm.matchedLen > 0:
      c.caseId = c.url.substr(mm.group(0,2).a, mm.group(0,2).b)
    if "/ie/cases/" in c.url: c.jurisdiction = "Ireland"
    elif "/ni/cases/" in c.url: c.jurisdiction = "Northern Ireland"
    elif "/scot/cases/" in c.url: c.jurisdiction = "Scotland"
    else: c.jurisdiction = self.jurisdiction()
    out.add c
    if out.len >= limit: break
  return out

proc getCaseById*(self: BAILIIScraper; caseId: string): CaseData =
  if caseId.len == 0: return initCaseData()
  var url = caseId
  if not url.startsWith("http"):
    if url.startsWith("uk/cases/") or url.startsWith("ie/cases/") or url.startsWith("ni/cases/") or url.startsWith("scot/cases/"):
      url = self.baseUrl() & "/" & url & ".html"
  let resp = self.makeRequest(url)
  var c = initCaseData()
  c.url = url
  let titleRe = re"<title>(.*?)</title>"; let mm = titleRe.find(resp.body)
  if mm.matchedLen > 0: c.caseName = sanitizeText(resp.body.substr(mm.group(0,1).a, mm.group(0,1).b))

  # court patterns
  let courtPatterns = [re"(Supreme Court|Court of Appeal|High Court|Crown Court|Magistrates|Employment Tribunal)", re"(UKSC|EWCA|EWHC|UKUT|UKFTT)", re"(House of Lords|Privy Council)"]
  for cp in courtPatterns:
    let cm = cp.find(resp.body)
    if cm.matchedLen > 0:
      c.court = normalizeCourtName(sanitizeText(resp.body.substr(cm.group(0,1).a, cm.group(0,1).b)))
      break

  # date patterns
  let datePatterns = [re"(\d{1,2}\s+(?:January|February|March|April|May|June|July|August|September|October|November|December)\s+\d{4})", re"(\d{4}-\d{2}-\d{2})", re"(\d{1,2}/\d{1,2}/\d{4})"]
  for dp in datePatterns:
    let dm = dp.find(resp.body)
    if dm.matchedLen > 0:
      let dateStr = resp.body.substr(dm.group(0,1).a, dm.group(0,1).b)
      try:
        if dateStr.contains('-'): c.date = parse("yyyy-MM-dd", dateStr)
        elif dateStr.contains('/'): c.date = parse("dd/MM/yyyy", dateStr)
        else: c.date = parse("d MMMM yyyy", dateStr)
      except: discard
      break

  # citations
  for cm2 in findAll(resp.body, re"\[(\d{4})\]\s+(UKSC|EWCA|EWHC|UKUT)\s+(\d+)"):
    let y = resp.body.substr(cm2.group(0,1).a, cm2.group(0,1).b)
    let courtCode = resp.body.substr(cm2.group(0,2).a, cm2.group(0,2).b)
    let num = resp.body.substr(cm2.group(0,3).a, cm2.group(0,3).b)
    c.citations.add "[" & y & "] " & courtCode & " " & num

  # full text
  let contentRe = re"<div[^>]*(?:class=\"(judgment|content)\"|id=\"main\")[^>]*>([\s\S]*?)</div>"
  let contm = contentRe.find(resp.body)
  if contm.matchedLen > 0: c.fullText = sanitizeText(resp.body.substr(contm.group(0,2).a, contm.group(0,2).b))

  # judges
  if c.fullText.len > 0:
    for jm in findAll(c.fullText, re"(?:Lord|Lady|Mr|Mrs|Ms)\s+Justice\s+([A-Z][a-z]+(?:\s+[A-Z][a-z]+)*)|([A-Z][a-z]+\s+LJ)|([A-Z][a-z]+\s+J\.?)"):
      let seg = c.fullText.substr(jm.group(0,0).a, jm.group(0,0).b)
      var name = seg.replace(" LJ", "").replace(" J.", "").replace(" J", "")
      name = name.replace(re"^(?:Lord|Lady|Mr|Mrs|Ms)\s+Justice\s+", "")
      if name.len > 0 and c.judges.len < 5 and not (name in c.judges): c.judges.add name

  # id & jurisdiction from URL
  let idm = re"/(uk|ie|ni|scot)/cases/([^/]+/\d+/\d+)".find(url)
  if idm.matchedLen > 0:
    c.caseId = url.substr(idm.group(0,2).a, idm.group(0,2).b)
    let jur = url.substr(idm.group(0,1).a, idm.group(0,1).b)
    c.jurisdiction = (if jur == "ie": "Ireland" elif jur == "ni": "Northern Ireland" elif jur == "scot": "Scotland" else: self.jurisdiction())
  return c

File: internal/scraper/jurisdictions/bailii.go
package jurisdictions

import (
	"context"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/PuerkitoBio/goquery"
	"github.com/gongahkia/kite/internal/scraper"
	"github.com/gongahkia/kite/pkg/errors"
	"github.com/gongahkia/kite/pkg/models"
)

// BAILIIScraper implements scraping for BAILII (British and Irish Legal Information Institute)
type BAILIIScraper struct {
	*scraper.BaseScraper
	baseURL string
	client  *http.Client
}

// NewBAILIIScraper creates a new BAILII scraper
func NewBAILIIScraper() *BAILIIScraper {
	baseURL := "https://www.bailii.org"
	base := scraper.NewBaseScraper(
		"BAILII",
		"United Kingdom & Ireland",
		baseURL,
		12, // 12 requests per minute to be conservative
	)

	return &BAILIIScraper{
		BaseScraper: base,
		baseURL:     baseURL,
		client: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// SearchCases searches for cases matching the query
func (bs *BAILIIScraper) SearchCases(ctx context.Context, query scraper.SearchQuery) ([]*models.Case, error) {
	// Build search URL
	searchURL, err := bs.buildSearchURL(query)
	if err != nil {
		return nil, errors.ParsingError("failed to build search URL", err)
	}

	// Check robots.txt
	allowed, err := bs.BaseScraper.client.CheckRobots(ctx, "/form/search_multidatabase.html")
	if err != nil || !allowed {
		return nil, errors.ErrRobotsDisallowed
	}

	// Wait for rate limit
	if err := bs.BaseScraper.client.rateLimiter.Wait(ctx); err != nil {
		return nil, errors.RateLimitError("rate limit exceeded")
	}

	// Make HTTP request
	req, err := http.NewRequestWithContext(ctx, "GET", searchURL, nil)
	if err != nil {
		return nil, errors.NetworkError("failed to create request", err)
	}

	req.Header.Set("User-Agent", bs.BaseScraper.client.userAgent)

	resp, err := bs.client.Do(req)
	if err != nil {
		return nil, errors.NetworkError("failed to fetch search results", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, errors.NetworkError(fmt.Sprintf("unexpected status code: %d", resp.StatusCode), nil)
	}

	// Parse HTML
	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, errors.ParsingError("failed to parse HTML", err)
	}

	// Extract cases from search results
	cases := make([]*models.Case, 0)

	doc.Find("li.resultItem").Each(func(i int, s *goquery.Selection) {
		if query.Limit > 0 && len(cases) >= query.Limit {
			return
		}

		caseData := bs.extractCaseFromSearchResult(s)
		if caseData != nil {
			cases = append(cases, caseData)
		}
	})

	return cases, nil
}

// GetCaseByID retrieves a specific case by its ID (e.g., "UKSC/2023/15")
func (bs *BAILIIScraper) GetCaseByID(ctx context.Context, caseID string) (*models.Case, error) {
	// BAILII case IDs are typically in format: UKSC/2023/15
	// Build case URL from ID
	caseURL := bs.buildCaseURL(caseID)

	// Check robots.txt
	allowed, err := bs.BaseScraper.client.CheckRobots(ctx, "/")
	if err != nil || !allowed {
		return nil, errors.ErrRobotsDisallowed
	}

	// Wait for rate limit
	if err := bs.BaseScraper.client.rateLimiter.Wait(ctx); err != nil {
		return nil, errors.RateLimitError("rate limit exceeded")
	}

	// Make HTTP request
	req, err := http.NewRequestWithContext(ctx, "GET", caseURL, nil)
	if err != nil {
		return nil, errors.NetworkError("failed to create request", err)
	}

	req.Header.Set("User-Agent", bs.BaseScraper.client.userAgent)

	resp, err := bs.client.Do(req)
	if err != nil {
		return nil, errors.NetworkError("failed to fetch case", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		return nil, errors.ErrNotFound
	}

	if resp.StatusCode != http.StatusOK {
		return nil, errors.NetworkError(fmt.Sprintf("unexpected status code: %d", resp.StatusCode), nil)
	}

	// Parse HTML
	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, errors.ParsingError("failed to parse HTML", err)
	}

	// Extract case details
	return bs.extractCaseDetails(doc, caseID, caseURL)
}

// GetCasesByDateRange retrieves cases within a date range
func (bs *BAILIIScraper) GetCasesByDateRange(ctx context.Context, startDate, endDate time.Time, limit int) ([]*models.Case, error) {
	query := scraper.SearchQuery{
		StartDate: &startDate,
		EndDate:   &endDate,
		Limit:     limit,
	}

	return bs.SearchCases(ctx, query)
}

// IsAvailable checks if BAILII is available
func (bs *BAILIIScraper) IsAvailable(ctx context.Context) bool {
	req, err := http.NewRequestWithContext(ctx, "HEAD", bs.baseURL, nil)
	if err != nil {
		return false
	}

	resp, err := bs.client.Do(req)
	if err != nil {
		return false
	}
	defer resp.Body.Close()

	return resp.StatusCode == http.StatusOK
}

// buildSearchURL builds the search URL with query parameters
func (bs *BAILIIScraper) buildSearchURL(query scraper.SearchQuery) (string, error) {
	params := url.Values{}

	if query.Query != "" {
		params.Set("query", query.Query)
	}

	if query.Court != "" {
		params.Set("court", query.Court)
	}

	if query.StartDate != nil && query.EndDate != nil {
		// BAILII uses a date range format
		dateRange := fmt.Sprintf("%s to %s",
			query.StartDate.Format("02/01/2006"),
			query.EndDate.Format("02/01/2006"))
		params.Set("date", dateRange)
	}

	// Set default database to search all UK and Irish databases
	params.Set("database", "all")
	params.Set("method", "boolean")

	searchURL := fmt.Sprintf("%s/cgi-bin/lucy_search_1.cgi?%s", bs.baseURL, params.Encode())
	return searchURL, nil
}

// buildCaseURL builds a case URL from a case citation (e.g., "UKSC/2023/15")
func (bs *BAILIIScraper) buildCaseURL(caseID string) string {
	// BAILII URLs are structured as: /jurisdiction/court/year/number.html
	// Example: /uk/cases/UKSC/2023/15.html

	// Clean the caseID
	caseID = strings.TrimSpace(caseID)

	// If caseID already looks like a path, use it
	if strings.Contains(caseID, "/") {
		// Extract components
		parts := strings.Split(caseID, "/")
		if len(parts) >= 3 {
			// Format: court/year/number
			court := strings.ToUpper(parts[0])
			year := parts[1]
			number := parts[2]

			// Determine jurisdiction from court code
			jurisdiction := "uk"
			if strings.Contains(court, "IE") || strings.Contains(court, "IESC") {
				jurisdiction = "ie"
			}

			return fmt.Sprintf("%s/%s/cases/%s/%s/%s.html",
				bs.baseURL, jurisdiction, court, year, number)
		}
	}

	// Default: assume UK Supreme Court format
	return fmt.Sprintf("%s/uk/cases/%s.html", bs.baseURL, caseID)
}

// extractCaseFromSearchResult extracts case data from a search result item
func (bs *BAILIIScraper) extractCaseFromSearchResult(s *goquery.Selection) *models.Case {
	c := models.NewCase()

	// Extract case name and URL
	titleLink := s.Find("a.resultTitle")
	caseName := titleLink.Text()
	c.CaseName = strings.TrimSpace(caseName)

	// Extract case URL
	caseURL, exists := titleLink.Attr("href")
	if exists {
		if !strings.HasPrefix(caseURL, "http") {
			caseURL = bs.baseURL + caseURL
		}
		c.URL = caseURL

		// Extract ID from URL
		// Example: /uk/cases/UKSC/2023/15.html -> UKSC/2023/15
		if strings.Contains(caseURL, "/cases/") {
			parts := strings.Split(caseURL, "/cases/")
			if len(parts) >= 2 {
				id := strings.TrimSuffix(parts[1], ".html")
				c.ID = id
			}
		}
	}

	// Extract citation
	citation := s.Find(".resultCitation").Text()
	c.CaseNumber = strings.TrimSpace(citation)

	// Extract court from metadata
	court := s.Find(".resultCourt").Text()
	c.Court = strings.TrimSpace(court)

	// Extract date from metadata
	dateStr := s.Find(".resultDate").Text()
	if dateStr != "" {
		// Try to parse various date formats used by BAILII
		formats := []string{
			"02 January 2006",
			"2 January 2006",
			"02/01/2006",
			"2006-01-02",
		}
		for _, format := range formats {
			if date, err := time.Parse(format, strings.TrimSpace(dateStr)); err == nil {
				c.DecisionDate = &date
				break
			}
		}
	}

	// Extract snippet/summary
	snippet := s.Find(".resultSnippet").Text()
	c.Summary = strings.TrimSpace(snippet)

	// Determine jurisdiction from URL or court
	if strings.Contains(c.URL, "/ie/") || strings.Contains(c.Court, "Irish") {
		c.Jurisdiction = "Ireland"
	} else {
		c.Jurisdiction = "United Kingdom"
	}

	// Set metadata
	c.SourceDatabase = "BAILII"
	c.ScrapedAt = time.Now()
	c.LastUpdated = time.Now()
	c.Language = "en"
	c.Status = models.CaseStatusActive

	return c
}

// extractCaseDetails extracts detailed case information from a case page
func (bs *BAILIIScraper) extractCaseDetails(doc *goquery.Document, caseID, caseURL string) (*models.Case, error) {
	c := models.NewCase()

	c.ID = caseID
	c.URL = caseURL

	// Extract case name - BAILII uses various selectors
	caseName := doc.Find("h1.case-title").First().Text()
	if caseName == "" {
		caseName = doc.Find("blockquote > p > b").First().Text()
	}
	if caseName == "" {
		caseName = doc.Find("h2").First().Text()
	}
	c.CaseName = strings.TrimSpace(caseName)

	// Extract neutral citation
	citation := doc.Find(".citation").Text()
	if citation == "" {
		// Try alternative selectors
		citation = doc.Find("p:contains('[')").First().Text()
	}
	c.CaseNumber = strings.TrimSpace(citation)

	// Extract court
	court := doc.Find(".court").Text()
	if court == "" {
		// Extract from citation/URL
		if strings.Contains(caseID, "UKSC") {
			court = "UK Supreme Court"
		} else if strings.Contains(caseID, "EWCA") {
			court = "Court of Appeal (England & Wales)"
		} else if strings.Contains(caseID, "EWHC") {
			court = "High Court (England & Wales)"
		}
	}
	c.Court = strings.TrimSpace(court)

	// Extract date
	dateStr := doc.Find(".judgment-date").Text()
	if dateStr == "" {
		// Try to find date in various formats
		doc.Find("p").Each(func(i int, s *goquery.Selection) {
			text := s.Text()
			if strings.Contains(text, "Date:") || strings.Contains(text, "Judgment Date:") {
				dateStr = text
				return
			}
		})
	}

	if dateStr != "" {
		// Clean up date string
		dateStr = strings.TrimSpace(dateStr)
		dateStr = strings.TrimPrefix(dateStr, "Date:")
		dateStr = strings.TrimPrefix(dateStr, "Judgment Date:")
		dateStr = strings.TrimSpace(dateStr)

		// Try to parse various date formats
		formats := []string{
			"02 January 2006",
			"2 January 2006",
			"02/01/2006",
			"2006-01-02",
			"January 2, 2006",
		}
		for _, format := range formats {
			if date, err := time.Parse(format, dateStr); err == nil {
				c.DecisionDate = &date
				break
			}
		}
	}

	// Extract judges
	doc.Find(".judge").Each(func(i int, s *goquery.Selection) {
		judge := strings.TrimSpace(s.Text())
		if judge != "" {
			c.Judges = append(c.Judges, judge)
		}
	})

	// If no judges found with .judge selector, try alternative
	if len(c.Judges) == 0 {
		doc.Find("p:contains('Before:')").Each(func(i int, s *goquery.Selection) {
			text := s.Text()
			// Extract judge names after "Before:"
			if strings.Contains(text, "Before:") {
				parts := strings.Split(text, "Before:")
				if len(parts) > 1 {
					judgeText := parts[1]
					// Split by common separators
					judges := strings.Split(judgeText, ",")
					for _, judge := range judges {
						judge = strings.TrimSpace(judge)
						if judge != "" {
							c.Judges = append(c.Judges, judge)
						}
					}
				}
			}
		})
	}

	// Extract docket/case number
	docket := doc.Find(".docket-number").Text()
	if docket == "" {
		doc.Find("p").Each(func(i int, s *goquery.Selection) {
			text := s.Text()
			if strings.Contains(text, "Case No:") || strings.Contains(text, "Case Number:") {
				docket = text
				return
			}
		})
	}
	c.Docket = strings.TrimSpace(docket)

	// Extract full judgment text
	fullText := doc.Find(".judgment-body").Text()
	if fullText == "" {
		// Try alternative selectors
		fullText = doc.Find("ol[type='1']").Text()
		if fullText == "" {
			fullText = doc.Find("blockquote").Text()
		}
	}
	c.FullText = strings.TrimSpace(fullText)

	// Determine jurisdiction from URL
	if strings.Contains(caseURL, "/ie/") {
		c.Jurisdiction = "Ireland"
	} else {
		c.Jurisdiction = "United Kingdom"
	}

	// Set metadata
	c.SourceDatabase = "BAILII"
	c.ScrapedAt = time.Now()
	c.LastUpdated = time.Now()
	c.Language = "en"
	c.Status = models.CaseStatusActive

	return c, nil
}
